### PR TITLE
feat(): vpnkeyrotation apis, webhook, ITs, UTs and service layer

### PR DESCRIPTION
**Pull request by rahulsawra98**
# Description

This PR introduces new CRD APIs - vpnkeyrotation , webhook changes and UTs
Fixes #

## How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

## Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR requires documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code.
* [ ] I have commented my code, particularly in hard-to-understand areas.
* [ ] I have tested it for all user roles.
* [x] I have added all the required unit test cases.

## Does this PR introduce a breaking change for other components like worker-operator?

```release-note

```

### CODE REVIEW

File: apis/worker/v1alpha1/groupversion_info.go [hunk 1]
```diff
@@ -15,8 +15,8 @@
  */
 
 // Package v1alpha1 contains API Schema definitions for the worker v1alpha1 API group
-//+kubebuilder:object:generate=true
-//+groupName=worker.kubeslice.io
+// +kubebuilder:object:generate=true
```
Comment (rajendra-avesha) [inline apis/worker/v1alpha1/groupversion_info.go:18]: Is this change of extra space required?

File: config/crd/bases/controller.kubeslice.io_vpnkeyrotations.yaml [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+                additionalProperties:
+                  items:
+                    type: string
+                  type: array
+                description: ClusterGatewayMapping represents a map where key is cluster
+                  name and value is array of gateways present on that cluster. This
+                  is used to avoid unnecessary reconciliation in worker-operator.
+                type: object
+              sliceName:
+                description: 'INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
```
Comment (rajendra-avesha) [inline config/crd/bases/controller.kubeslice.io_vpnkeyrotations.yaml:59]: description should hold the slice name description

File: config/crd/bases/controller.kubeslice.io_vpnkeyrotations.yaml [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+                    lastUpdatedTimestamp:
+                      format: date-time
+                      type: string
+                    status:
+                      type: string
+                  required:
+                  - lastUpdatedTimestamp
+                  - status
+                  type: object
+                description: 'INSERT ADDITIONAL STATUS FIELD - define observed state
```
Comment (rajendra-avesha) [inline config/crd/bases/controller.kubeslice.io_vpnkeyrotations.yaml:78]: Can we change the description

File: service/slice_config_webhook_validation.go [hunk 4]
```diff
@@ -107,6 +112,43 @@ func ValidateSliceConfigDelete(ctx context.Context, sliceConfig *controllerv1alp
...
 }
 
+func validateCertsRotationInterval(ctx context.Context, sliceConfig *controllerv1alpha1.SliceConfig, old runtime.Object) *field.Error {
+	oldSliceConfig := old.(*controllerv1alpha1.SliceConfig)
+	// nochange detected
+	if sliceConfig.Spec.RenewBefore == oldSliceConfig.Spec.RenewBefore || sliceConfig.Spec.RenewBefore.IsZero() {
+		return nil
+	}
+	// change detected
+	vpnkeyrotaion := controllerv1alpha1.VpnKeyRotation{}
```
Comment (rajendra-avesha) [inline service/slice_config_webhook_validation.go:122]: Spell check of vpnkeyrotaion
Please make camel case

File: service/slice_config_webhook_validation.go [hunk 5]
```diff
@@ -107,6 +112,43 @@ func ValidateSliceConfigDelete(ctx context.Context, sliceConfig *controllerv1alp
...
+	exists, _ := util.GetResourceIfExist(ctx, types.NamespacedName{
+		Namespace: sliceConfig.Namespace,
+		Name:      sliceConfig.Name,
+	}, &vpnkeyrotaion)
+	if exists {
+		for cluster := range vpnkeyrotaion.Status.CurrentRotationState {
```
Comment (rajendra-avesha) [inline service/slice_config_webhook_validation.go:128]: I think vpnkeyrotaion.Status.CurrentRotationState shall give the map of gateways right not clusters ?

File: service/slice_config_webhook_validation.go [hunk 6]
```diff
@@ -107,6 +112,43 @@ func ValidateSliceConfigDelete(ctx context.Context, sliceConfig *controllerv1alp
 	return nil
 }
 
+func validateCertsRotationInterval(ctx context.Context, sliceConfig *controllerv1alpha1.SliceConfig, old runtime.Object) *field.Error {
```
Comment (rajendra-avesha) [inline service/slice_config_webhook_validation.go:115]: I think this API does validation of renewNow?

File: service/slice_config_webhook_validation.go [hunk 7]
```diff
@@ -107,6 +112,43 @@ func ValidateSliceConfigDelete(ctx context.Context, sliceConfig *controllerv1alp
...
+	}, &vpnkeyrotaion)
+	if exists {
+		for cluster := range vpnkeyrotaion.Status.CurrentRotationState {
+			status, ok := vpnkeyrotaion.Status.CurrentRotationState[cluster]
+			if ok {
+				if status.Status != controllerv1alpha1.Complete {
+					return &field.Error{
+						Type:   field.ErrorTypeForbidden,
+						Field:  "Field: RenewBefore",
+						Detail: fmt.Sprintf("Certs Renewal status for %s cluster is not in Complete state", cluster),
```
Comment (rajendra-avesha) [inline service/slice_config_webhook_validation.go:135]: I feel the detail be Key Rotation is in Progress

File: service/slice_config_webhook_validation_test.go [hunk 8]
```diff
@@ -1420,7 +1426,135 @@ func ValidateMaxClusterForParticipatingCluster(t *testing.T) {
 	require.Contains(t, err.Error(), "participating clusters cannot be greater than MaxClusterCount")
 	clientMock.AssertExpectations(t)
 }
+func TestValidateCertsRotationInterval_Posstive(t *testing.T) {
```
Comment (rajendra-avesha) [inline service/slice_config_webhook_validation_test.go:1429]: spell check TestValidateCertsRotationInterval_Posstive Positive

File: service/slice_config_webhook_validation_test.go [hunk 9]
```diff
@@ -1420,7 +1426,135 @@ func ValidateMaxClusterForParticipatingCluster(t *testing.T) {
...
+			},
+		}
+	}).Once()
+	oldSliceConfig := controllerv1alpha1.SliceConfig{}
+	err := validateCertsRotationInterval(ctx, sliceConfig, &oldSliceConfig)
+	require.NotNil(t, err)
+	require.Equal(t, err.Type, field.ErrorTypeForbidden)
+}
+
+func TestValidateCertsRotationInterval_PossitiveClusterStatus(t *testing.T) {
```
Comment (rajendra-avesha) [inline service/slice_config_webhook_validation_test.go:1518]: spell check for TestValidateCertsRotationInterval_PossitiveClusterStatus positive

**Review by rajendra-avesha — CHANGES_REQUESTED** [review]

**Review by rajendra-avesha — CHANGES_REQUESTED** [review]
Please check comments provided

File: service/slice_config_webhook_validation.go (hunk 5 above)
Comment (rahulsawra98) [reply service/slice_config_webhook_validation.go]: right!

File: service/slice_config_webhook_validation.go [hunk 10]
```diff
@@ -107,6 +112,43 @@ func ValidateSliceConfigDelete(ctx context.Context, sliceConfig *controllerv1alp
 	return nil
 }
 
+func validateRenewNowInSliceConfig(ctx context.Context, sliceConfig *controllerv1alpha1.SliceConfig, old runtime.Object) *field.Error {
```
Comment (narmidm) [inline service/slice_config_webhook_validation.go:115]: pointer changes.

File: service/slice_config_webhook_validation.go [hunk 11]
```diff
@@ -107,6 +112,43 @@ func ValidateSliceConfigDelete(ctx context.Context, sliceConfig *controllerv1alp
...
+	// check if we are past and its a correct time
+	if time.Now().After(sliceConfig.Spec.RenewBefore.Time) {
+		return nil
+	}
+	return &field.Error{
+		Type:   field.ErrorTypeForbidden,
+		Field:  "Field: RenewBefore",
+		Detail: "Renewal Time inappropriate for sliceconfig",
+	}
+}
```
Comment (narmidm) [inline service/slice_config_webhook_validation.go:150]: end statement with nil

File: service/slice_config_webhook_validation_test.go [hunk 12]
```diff
@@ -1420,7 +1426,135 @@ func ValidateMaxClusterForParticipatingCluster(t *testing.T) {
 	require.Contains(t, err.Error(), "participating clusters cannot be greater than MaxClusterCount")
 	clientMock.AssertExpectations(t)
 }
+func TestValidateCertsRotationInterval_Postive(t *testing.T) {
```
Comment (rajendra-avesha) [inline service/slice_config_webhook_validation_test.go:1429]: Earlier I had given comments to fix the spelling of postive TestValidateCertsRotationInterval_Postive?

File: service/slice_config_webhook_validation_test.go [hunk 13]
```diff
@@ -1420,7 +1426,135 @@ func ValidateMaxClusterForParticipatingCluster(t *testing.T) {
...
+			CertificateExpiryTime:   metav1.Time{Time: expiry},
+		}
+	}).Once()
+	oldSliceConfig := controllerv1alpha1.SliceConfig{}
+	err := validateRenewNowInSliceConfig(ctx, sliceConfig, &oldSliceConfig)
+	require.NotNil(t, err)
+}
+
+func TestValidateCertsRotationInterval_NegativeClusterStatus(t *testing.T) {
+	name := "slice_config"
```
Comment (rajendra-avesha) [inline service/slice_config_webhook_validation_test.go:1477]: This story was not closed as the ITs were not created in the last two sprints but still we have the same issue
We need have the ITs to cover the Key Rotation Interval checks

**Review by rajendra-avesha — CHANGES_REQUESTED** [review]

File: service/vpn_key_rotation_service.go [hunk 14]
```diff
@@ -0,0 +1,305 @@
...
+	listOpts := []client.ListOption{
+		client.MatchingLabels(
+			o,
+		),
+	}
+	if err := util.ListResources(ctx, &jobs, listOpts...); err != nil {
+		return false
+	}
+	for _, job := range jobs.Items {
+		for _, condition := range job.Status.Conditions {
```
Comment (rajendra-avesha) [inline service/vpn_key_rotation_service.go:289]: Please add TTL mechanism for finished Jobs [https://kubernetes.io/docs/concepts/workloads/controllers/job/#ttl-mechanism-for-finished-jobs] so that old job doesn't stagnate

File: service/vpn_key_rotation_service.go [hunk 15]
```diff
@@ -0,0 +1,305 @@
...
+			if !(condition.Type == batchv1.JobComplete && condition.Status == corev1.ConditionTrue) {
+				return false
+			}
+		}
```
Comment (rajendra-avesha) [inline service/vpn_key_rotation_service.go:293]: What if the JobConditionType  is 	Suspended or Failed?
I think we need to handle the case so that we don't be in the infinite loop

File: service/worker_slice_gateway_service.go [hunk 16]
```diff
@@ -627,7 +624,7 @@ func (s *WorkerSliceGatewayService) generateCerts(ctx context.Context, sliceName
 	environment["SLICE_NAME"] = sliceName
 	environment["CERT_GEN_REQUESTS"], _ = util.EncodeToBase64(&cpr)
 	util.CtxLogger(ctx).Info("jobNamespace", jobNamespace) //todo:remove
-	_, err := s.js.CreateJob(ctx, jobNamespace, JobImage, environment)
+	_, err := s.js.CreateJob(ctx, os.Getenv("KUBESLICE_CONTROLLER_MANAGER_NAMESPACE"), JobImage, environment)
```
Comment (rajendra-avesha) [inline service/worker_slice_gateway_service.go:627]: can you please take care of 	util.CtxLogger(ctx).Info("jobNamespace", jobNamespace) //todo:remove

File: util/common.go [hunk 17]
```diff
@@ -28,6 +28,16 @@ import (
 	"go.uber.org/zap/zapcore"
 )
 
+type WorkerSliceGatewayNetworkAddresses struct {
```
Comment (rajendra-avesha) [inline util/common.go:31]: Why did you move WorkerSliceGatewayNetworkAddresses to utils/common This doesn't seem to best practice

**Review by rajendra-avesha — CHANGES_REQUESTED** [review]

File: apis/controller/v1alpha1/vpnkeyrotation_webhook.go [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+	sliceconfigurationlog.Info("validate create", "name", r.Name)
+	sliceConfigCtx := util.PrepareKubeSliceControllersRequestContext(context.Background(), vpnKeyRoationConfigWebhookClient, nil, "VpnKeyRotationConfigValidation", nil)
+	return customVpnKeyRotationCreateValidation(sliceConfigCtx, r)
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (r *VpnKeyRotation) ValidateUpdate(old runtime.Object) error {
+	vpnkeyrotationlog.Info("validate update", "name", r.Name)
+
+	// TODO(user): fill in your validation logic upon object update.
```
[the remark this answers is not in the document]
Comment (rahulsawra98) [reply apis/controller/v1alpha1/vpnkeyrotation_webhook.go:62]: FYI: This is a vpn key rotation webhook - part of next sprint!

File: util/common.go (hunk 17 above)
Comment (rahulsawra98) [reply util/common.go:31]: Yeah it has to be moved back!

File: controllers/controller/vpnkey_rotation_controller_test.go [hunk 19]
```diff
@@ -0,0 +1,249 @@
...
+			getKey := types.NamespacedName{
+				Namespace: vpn.Namespace,
+				Name:      vpn.Name,
+			}
+			// check if creation TS is not zero
+			Eventually(func() bool {
+				err := k8sClient.Get(ctx, getKey, createdVpnKeyConfig)
+				Expect(err).To(BeNil())
+				return !createdVpnKeyConfig.Spec.CertificateCreationTime.IsZero()
+			}, timeout, interval).Should(BeTrue())
```
[the remark this answers is not in the document]
Comment (rahulsawra98) [reply controllers/controller/vpnkey_rotation_controller_test.go:158]: Don;t think it matters because all we care about is if it has updated the CreationTS and ExpiryTS

File: service/vpn_key_rotation_service.go [hunk 20]
```diff
@@ -0,0 +1,305 @@
...
+		return ctrl.Result{}, err
+	}
+	copyVpnConfig := vpnKeyRotationConfig.DeepCopy()
+
+	if !reflect.DeepEqual(copyVpnConfig.Spec.ClusterGatewayMapping, clusterGatewayMapping) {
+		copyVpnConfig.Spec.ClusterGatewayMapping = clusterGatewayMapping
+		if err := util.UpdateResource(ctx, copyVpnConfig); err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{Requeue: true}, nil
```
[the remark this answers is not in the document]
Comment (rahulsawra98) [reply service/vpn_key_rotation_service.go:139]: As explained!

File: service/vpn_key_rotation_service.go [hunk 21]
```diff
@@ -0,0 +1,305 @@
...
+		if err := util.UpdateResource(ctx, copyVpnConfig); err != nil {
+			return nil, err
+		}
+	} else {
+		if now.After(copyVpnConfig.Spec.CertificateExpiryTime.Time) {
+			if err := v.triggerJobsForCertCreation(ctx, copyVpnConfig, s); err != nil {
+				logger.Error("error creating new certs", err)
+				return nil, err
+			}
+			copyVpnConfig.Spec.CertificateCreationTime = &now
```
[the remark this answers is not in the document]
Comment (rahulsawra98) [reply service/vpn_key_rotation_service.go:178]: Please check now!

File: service/vpn_key_rotation_service.go (hunk 14 above)
Comment (rahulsawra98) [reply service/vpn_key_rotation_service.go:300]: Its already set to 24 hours - 1 day.

File: apis/controller/v1alpha1/vpnkeyrotation_webhook.go [hunk 22]
```diff
@@ -0,0 +1,72 @@
...
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+// log is for logging in this package.
+var (
+	vpnkeyrotationlog                    = logf.Log.WithName("vpnkeyrotation-resource")
```
[the remark this answers is not in the document]
Comment (rahulsawra98) [reply apis/controller/v1alpha1/vpnkeyrotation_webhook.go:31]: okay will do but it was auto generated code from kubebuilder!

File: util/common.go (hunk 17 above)
Comment (rahulsawra98) [reply util/common.go:31]: Due to import cycle issues.
service package imports mock package.
mock package imports service package.

Hence it was moved here!

**Comment by NishantSingh10** [comment]
report link  'https://kubeslice.github.io/e2e-allure-reports/Kind-kubeslice-controller-2023-06-27T09:54:41-dev-163/index.html'

**Comment by NishantSingh10** [comment]
report link  'https://kubeslice.github.io/e2e-allure-reports/Kind-kubeslice-controller-2023-06-27T11:25:21-dev-164/index.html'

File: config/crd/bases/controller.kubeslice.io_vpnkeyrotations.yaml [hunk 23]
```diff
@@ -0,0 +1,108 @@
...
+                  all the gateway pairs will expire
+                format: date-time
+                type: string
+              clusters:
+                description: clusters contains the list of clusters attached to this
+                  slice
+                items:
+                  type: string
+                type: array
+              gateways:
```
[the remark this answers is not in the document]
Comment (rajendra-avesha) [reply config/crd/bases/controller.kubeslice.io_vpnkeyrotations.yaml:54]: Can you please check "gateways" is valid field I think it should be ClusterGatewayMapping please look at it

File: controllers/controller/vpnkey_rotation_controller_test.go [hunk 24]
```diff
@@ -0,0 +1,298 @@
...
+				Expect(err).To(BeNil())
+				return createdVpnKeyConfig.Spec.ClusterGatewayMapping != nil
+			}, timeout, interval).Should(BeTrue())
+
+			// the length of map should be 3
+			Eventually(func() bool {
+				err := k8sClient.Get(ctx, getKey, createdVpnKeyConfig)
+				Expect(err).To(BeNil())
+				return len(createdVpnKeyConfig.Spec.ClusterGatewayMapping) == 3
+			}, timeout, interval).Should(BeTrue())
```
Comment (rajendra-avesha) [inline controllers/controller/vpnkey_rotation_controller_test.go:248]: Add the validation for ClusterGatewayMapping map

File: service/vpn_key_rotation_service.go [hunk 25]
```diff
@@ -0,0 +1,305 @@
...
+	return clusterGatewayMapping, nil
+}
+
+func (v *VpnKeyRotationService) triggerJobsForCertCreation(ctx context.Context, vpnKeyRotationConfig *controllerv1alpha1.VpnKeyRotation, s *controllerv1alpha1.SliceConfig) error {
+	o := map[string]string{
+		"original-slice-name": vpnKeyRotationConfig.Spec.SliceName,
+	}
+	workerSliceGatewaysList, err := v.listWorkerSliceGateways(ctx, o)
+	if err != nil {
+		return err
```
[the remark this answers is not in the document]
Comment (rajendra-avesha) [reply service/vpn_key_rotation_service.go:252]: I still see error logs are not added in failure cases

File: service/vpn_key_rotation_service.go (hunk 15 above)
Comment (rajendra-avesha) [reply service/vpn_key_rotation_service.go:339]: Any update?

**Review by rajendra-avesha — CHANGES_REQUESTED** [review]

File: controllers/controller/vpnkey_rotation_controller_test.go [hunk 26]
```diff
@@ -0,0 +1,298 @@
...
+			Expect(k8sClient.Create(ctx, s)).Should(Not(Succeed()))
+
+			// RotationInterval < 30
+			s.Spec.RotationInterval = 20
+			// Expected Error:
+			// {
+			// 	Type: "FieldValueInvalid",
+			// 	Message: "Invalid value: 30: spec.rotationInterval in body should be greater than or equal to 30",
+			// 	Field: "spec.rotationInterval",
+			// },
```
[the remark this answers is not in the document]
Comment (rahulsawra98) [reply controllers/controller/vpnkey_rotation_controller_test.go:118]: How can you create events when api-server rejects the sliceconfig?

File: controllers/controller/vpnkey_rotation_controller_test.go [hunk 27]
```diff
@@ -0,0 +1,298 @@
...
+			s := slice.DeepCopy()
+			// RotationInterval > 90
+			// Expected Error:
+			//{
+			// 	Type: "FieldValueInvalid",
+			// 	Message: "Invalid value: 90: spec.rotationInterval in body should be less than or equal to 90",
+			// 	Field: "spec.rotationInterval",
+			// },
+			s.Spec.RotationInterval = 100
+			Expect(k8sClient.Create(ctx, s)).Should(Not(Succeed()))
```
[the remark this answers is not in the document]
Comment (rahulsawra98) [reply controllers/controller/vpnkey_rotation_controller_test.go:109]: Please check!

File: controllers/controller/vpnkey_rotation_controller_test.go (hunk 24 above)
Comment (rahulsawra98) [reply controllers/controller/vpnkey_rotation_controller_test.go:264]: Not sure what does that mean?

File: service/vpn_key_rotation_service.go [hunk 28]
```diff
@@ -0,0 +1,305 @@
...
+func (v *VpnKeyRotationService) ReconcileClusters(ctx context.Context, sliceName, namespace string, clusters []string) (*controllerv1alpha1.VpnKeyRotation, error) {
+	vpnKeyRotationConfig := controllerv1alpha1.VpnKeyRotation{}
+	found, err := util.GetResourceIfExist(ctx, types.NamespacedName{
+		Namespace: namespace,
+		Name:      sliceName,
+	}, &vpnKeyRotationConfig)
+	if err != nil {
+		return nil, err
+	}
+	if found {
```
[the remark this answers is not in the document]
Comment (rahulsawra98) [reply service/vpn_key_rotation_service.go:106]: handled with finalizer

File: service/vpn_key_rotation_service.go (hunk 15 above)
Comment (rahulsawra98) [reply service/vpn_key_rotation_service.go:339]: https://github.com/kubeslice/kubeslice-controller/pull/169

File: config/crd/bases/controller.kubeslice.io_vpnkeyrotations.yaml (hunk 23 above)
[the remark this answers is not in the document]
Comment (rahulsawra98) [reply config/crd/bases/controller.kubeslice.io_vpnkeyrotations.yaml:54]: done in the next PR!

**Comment by NishantSingh10** [comment]
report link  'https://kubeslice.github.io/e2e-allure-reports/Kind-kubeslice-controller-2023-06-29T08:52:33-dev-165/index.html'

**Review by rajendra-avesha — APPROVED** [review]

**Comment by NishantSingh10** [comment]
report link  'https://kubeslice.github.io/e2e-allure-reports/Kind-kubeslice-controller-2023-07-06T07:27:45-dev-166/index.html'

File: config/crd/kustomization.yaml [hunk 29]
```diff
@@ -11,6 +11,7 @@ resources:
   - bases/worker.kubeslice.io_workerserviceimports.yaml
   - bases/controller.kubeslice.io_sliceqosconfigs.yaml
   - bases/worker.kubeslice.io_workerslicegwrecyclers.yaml
+- bases/controller.kubeslice.io_vpnkeyrotations.yaml
```
Comment (priyank-upadhyay) [inline config/crd/kustomization.yaml:14]: It seems alignment is not correct.

File: go.mod
Comment (narmidm) [inline go.mod:1]: go 1.18 -> go 1.19

**Review by narmidm — APPROVED** [review]

**Review by narmidm — APPROVED** [review]

**Review by PuranAdhikari — APPROVED** [review]
